### ISSUE TITLE
CTRL+C does not work correctly

### ISSUE THREAD

**Issue by FeralChild64** [dosbox-staging/dosbox-staging#2044]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

0.80.0-alpha

### What Operating System are you using?

Linux x86_64

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

I am unable to terminate the original MORE command with CTRL+C.

Problem does not exist on DOSBox-X 222.09.01, it reacts like the original DOS.

### Steps to reproduce the behaviour.

1. Take MORE.COM from the original DOS (I have tried MS-DOS 6.22, PC-DOS 2000 and DR-DOS 7.03)
2. Launch it without any arguments (just type MORE in a directory the command is located)
3. Press CTRL+C - on real DOS it terminates the command and returns to prompt, on DOSBox Staging it does not (NOT OK) - but, strangely, it can be terminated with pressing CTRL and a sequence ZXCVBNM. The command from FreeDOS can't be terminated in any way.

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by ningshuohao** [comment]
In windows, the bug is the same.

**Comment by joncampbell123** [comment]
For CTRL+C, the DOS console or DOS kernel need to check the input for (in limited circumstances) ASCII byte value 3 (CTRL+C) and signal a specific DOS interrupt when it sees it that either goes to a default handler, or goes to a custom handler installed by the DOS program. http://www.ctyme.com/intr/rb-4110.htm

CTRL+Break is the same but requires the BIOS IRQ1 handler to detect that keystroke and call another specific interrupt which then signals DOS to then trigger the CTRL+C handler. http://www.ctyme.com/intr/rb-2442.htm

On PC-98, the "STOP" key has a similar function using INT 6. Yes, that's right... on PC-98, interrupt 6 could either mean an invalid opcode exception OR the user pressed the STOP key. :laughing:

**Comment by Burrito78** [comment]
@joncampbell123 So would you say that this is a bug in Staging?

**Comment by joncampbell123** [comment]
Yes. CTRL+C is supposed to break out of the program when it calls for input from certain points in INT 21h. As is Ctrl+Break. You can borrow the code from DOSBox-X to handle that if you like. It needs changes to INT 21h, INT 16h, and the callback.cpp source file.

**Comment by kcgen** [comment]
Thank you for for this exacting help, @joncampbell123

**Comment by Burrito78** [comment]
Maniac Mansion and Zak McKracken are exited using CTRL+C in real DOS. That never worked in original DOSBox, as explained here: https://www.vogons.org/viewtopic.php?f=33&t=31571&p=259335&hilit=maniac%20mansion%20quit%20dosbox#p259335

Can this issue be solved also using the methods described above maybe?

**Comment by johnnovak** [comment]
> Maniac Mansion and Zak McKracken are exited using CTRL+C in real DOS.

I think we can agree that's pretty shit though 😄

**Comment by Grounded0** [comment]
I remember powering off my PC as a very very young child as i didn't know how to exit Zak McKracken.

**Comment by Grounded0** [comment]
Adding game compatibility since at least Maniac Mansion and Zak McKracken are affected.

**Comment by Grounded0** [comment]
Adding DOS label since we're not implementing parts of MS-DOS functionality.

**Comment by Grounded0** [comment]
Definitely my top picks to fix in 0.83.0. 🙂

**Comment by johnnovak** [comment]
> Definitely my top picks to fix for 0.83.0. 🙂

Why is that so? We don't miss anything by not having this if the goal is playing games.

**Comment by Grounded0** [comment]
> Why is that so? We don't miss anything by not having this if the goal is playing games.

At least Maniac Mansion and Zak McKracken are affected.

**Comment by johnnovak** [comment]
> > Why is that so? We don't miss anything by not having this if the goal is playing games.
> 
> At least Maniac Mansion and Zak McKracken are affected.

Yeah but who cares. You can close the emulator when you're finished playing the game 🤷🏻

**Comment by Grounded0** [comment]
> Yeah but who cares. You can close the emulator when you're finished playing the game 🤷🏻

This is the reason why I'm spending my money at Costa del Sol instead of here from now on.

**Comment by Grounded0** [comment]
<img width="1392" alt="Screenshot 2024-08-17 at 21 51 33" src="https://github.com/user-attachments/assets/064dca73-9e00-4ebf-9d08-4c0f594f24c5">

**Comment by Grounded0** [comment]
<img width="796" alt="Screenshot 2024-08-17 at 22 19 01" src="https://github.com/user-attachments/assets/4dbff7f8-7a01-417b-96f6-b9086fe1c777">

**Comment by Grounded0** [comment]
<img width="756" alt="Screenshot 2024-08-17 at 22 39 17" src="https://github.com/user-attachments/assets/ee34695e-a487-4f19-b440-dedcb774c925">
<img width="756" alt="Screenshot 2024-08-17 at 22 39 44" src="https://github.com/user-attachments/assets/bfe734d4-609b-46f9-a6d3-270d498b9b33">

**Comment by FeralChild64** [comment]
@Grounded0 Look, I fully agree this should be implemented at some point. But, the problem is, our resources are VERY limited.

For example, I am currently reworking locale settings and synchronization with host OS - as this does not work correctly in many cases and upsets several users. The amount of work is significant - untangling some legacy spaghetti code, a lot of new code, some small clever algorithms, ensuring compatibility with several Linux desktops (Wayland complicated a lot here), plus I had to look at every single Windows/Linux/macOS keyboard layout and map it to the most suitable FreeDOS layout, by hand.
I have started several months ago - in the mean time I needed to take several breaks due to (sometimes VERY unpleasant) personal issues, I was interrupting my work to fix a nasty keyboard regression, or to implement some minor enhancements.
I just can't take another task, as I have already some other half-baked stuff in line, some investigation started - all left to clean up this locale settings and implement them (hopefully) the right way, without too many shortcuts.

And the rest of the team has other fascinating things to do - for example, I suppose the threaded audio took a lot of work and testing, and despite already bringing a huge improvement, it still did not reach the desired target state.

We just don't have time to implement everything at once...

**Comment by Grounded0** [comment]
Yeah I know. I'm just dumping the basic understanding here how it works.

**Comment by johnnovak** [comment]
As @FeralChild64 has much more elaborately expressed, nobody is against this @Grounded0  -- it's just low priority in the grand scheme of things as it affects a very small number of real-world use cases.

For the record, my "flagship" ticket for 0.83.0 is NukedSC55 integration (via a plugin system to get around the old MAME license GPL2 incompatibility issue). That will bring virtually perfect cross-platform SC-55 emulation to the masses without having to mess with virtual MIDI cables.

And as my time and energy is limited, I work on the most significant features affecting the most users first. The funny thing is that obscure niche issues are not necessarily easier or less time-consuming to fix than implementing major features.

We gotta spend our energies wisely and prioritise, plus let's not forget this is a hobby where we even want to enjoy ourselves 😄

**Comment by interloper98** [comment]
Whenever we get Ctrl+C, you bet I'll be periodically starting up Maniac Mansion and Zak McKracken just to Ctrl+C those suckers back out to the C prompt! :metal:  :smiling_imp:

**Comment by Grounded0** [comment]
You can avoid this issue by playing original Maniac Mansion (1987) on Weird Ed's computer in the Day of the Tentacle (1993):

<img width="1392" alt="Screenshot 2024-08-18 at 1 57 10" src="https://github.com/user-attachments/assets/3bdf787d-056d-4ba1-99a0-e179501fd66f">
<img width="1392" alt="Screenshot 2024-08-18 at 1 59 41" src="https://github.com/user-attachments/assets/1c513df2-4e63-4d1e-8e19-eca0f68c5901">

**Comment by johnnovak** [comment]
> You can avoid this issue by playing original Maniac Mansion (1987) on Weird Ed's computer in the Day of the Tentacle (1993):

But the most pression question: can you Ctrl+C to exit MM and stay in DoTT? 🤔 😆

**Comment by Grounded0** [comment]
They implemented Alt+X to exit MANIAC for this interpreter version.

**Comment by Grounded0** [comment]
Friend of the project Aaron Giles explaining how he imported `MANIAC` to Mac OS port of `DOTT` (we love this stuff):

https://macintoshgarden.org/games/day-of-the-tentacle

<img width="742" alt="Screenshot 2024-08-18 at 4 55 12" src="https://github.com/user-attachments/assets/385b6fb7-93bb-4d74-bb6c-9fb04a1fc7f1">

**Comment by johnnovak** [comment]
Pretty cool story! 

By the way, LucasArt's first adventure game ever was [Labyrinth](https://en.wikipedia.org/wiki/Labyrinth:_The_Computer_Game), not Maniac Mansion.

Source:
https://en.wikipedia.org/wiki/LucasArts_adventure_games

**Comment by Grounded0** [comment]
> Whenever we get Ctrl+C, you bet I'll be randomly launching Maniac Mansion and Zak McKracken just to Ctrl+C those suckers back out to the C prompt! 🤘 😈

This is much more useful on DOS than on Unix based systems since software glitches more often and you really need to break out of them when that happens since they're on the drivers seat on a non-concurrent OS like DOS. I just pulled two games as an example since everything is tested here on that basis.

**Comment by interloper98** [comment]
I will appreciate (and use!) this feature @Grounded0. 

Seeing @FeralChild64 opened this shows the teams wants this, which is quite the opposite compared to the, well, non-comital responses I've seen here from the DOSBox Team members:

https://www.vogons.org/viewtopic.php?t=12160

https://www.vogons.org/viewtopic.php?t=33687

https://www.vogons.org/viewtopic.php?t=34914

So I don't think there's anything to worry about; there's a light at the end of this tunnel! 😌

**Comment by johnnovak** [comment]
Yeah, like I said, whoever does this, it will get merged (given it doesn't massively overcomplicate anything, but I doubt it).

But it really has very limited applicability in games. The Zak and MM examples are outliers; most games override the handler and Ctrl+C will do absolutely nothing in most games. However, it's useful in non-gaming contexts for long-running batch files and maybe really old games did not nuke the Ctrl+C handler.

Many 90s games even nuked the Ctrl+Alt+Del handler and you had to reach or the reset switch 😄 I _might_ have done that in my own programs on occasion because it was the 1337 thing to do 😎 

Our former teammate Wengier summed it up quite succinctly:

<img width="1630" alt="image" src="https://github.com/user-attachments/assets/c32185b8-1bca-4dc4-b2a4-19dd13d14bd2">

**Comment by johnnovak** [comment]
> @joncampbell123 So would you say that this is a bug in Staging?

Also, not "bug" per se, but an unimplemented feature that DOS has.

**Comment by Grounded0** [comment]
Problem with Harekiet and qbix was that they wouldn't had checked this in even if you provided a pristine implementation for them. And around a million other examples of this same insanity.

**Comment by johnnovak** [comment]
> The problem with Harekiet and qbix was that they wouldn't had checked this in even if you provided a pristine implementation for them. And around a million other examples of this same mentality.

Yeah, things don't work like that here 😏

**Comment by Dreamweb79** [comment]
In Flight Assignment - A.T.P. (every version of the game) and also in Airline Simulator (same codebase), there is a never resolved bug that make the game unresponsive to any user input, even if the game itself doesn't lockup.
This bug happens many times when you open the Flight Planning menu with the F7 key.
It's easy to reproduce, just continue to open/close that menu until it happen.
The only accepted input is ctrl-break, that open a popup where you can select Exit game/Continue Flight.
Selecting the last option fix the unresponsiveness.

**Comment by johnnovak** [comment]
> In Flight Assignment - A.T.P. (every version of the game) and also in Airline Simulator (same codebase), there is a never resolved bug that make the game unresponsive to any user input, even if the game itself doesn't lockup. This bug happens many times when you open the Flight Planning menu with the F7 key. It's easy to reproduce, just continue to open/close that menu until it happen. The only accepted input is ctrl-break, that open a popup where you can select Exit game/Continue Flight. Selecting the last option fix the unresponsiveness.

I don't see how this is related to the issue ticket. Please open separate issue ticket for vaguely related things. But... I'm not even sure this in an issue; you're just mentioning a workaround?

**Comment by Dreamweb79** [comment]
Yes maybe I could open a new issue for this, but I thought to put it here because I think it's related with the OP, where ctrl-c is often interchangeable with ctrl-break.
So reading through the comments about how useful this functionality could or not be for Maniac Mansion or Zak McKracken I wanted to add that for ATP and Airline Simulator the (ctrl-break functionality in this case) it's not a pointless requirements.
It's a workaround because it allows you to bypass the game bug (as in the real hardware) but it also open a menu of the game.

**Comment by johnnovak** [comment]
> It's a workaround because it allows you to bypass the game bug (as in the real hardware) but it also open a menu of the game.

Thanks for clarifying. It wasn't clear from your first comment that this is a game bug that happens on hardware too.

Yeah, in that case it makes sense, this is related to the ticket, indeed. Thanks for the info.

**Comment by ltning** [comment]
Hola, just chiming in with another use case for this: I'm developing software under DOS (CGI handler for the old SIOUX web server for DOS), and testing is *really* hard when you can't ctrl-break your way back to the editor/IDE or debugger.

I know, I know .. not a game, just putting it out there :)

**Comment by Grounded0** [comment]
I've got two more instances where Microsoft Flight Simulator 5.1 and and MyZ80 (Z80 emulator for CP/M) are broken due to not being able to access their Control+Break customised interrupt handlers.

MyZ80 can be downloaded from here: http://www.z80.eu/downloads/myz80124.zip

Some of these games have weird key combos to exit them I often forget so it would be much easier to hit that sweet Control+Break generic interrupt handler to exit them. Similar to using our own internal pause than trying to remember what weird key combo pauses this specific game.

This is probably our biggest DOS compatibility issue now after adding file locking support.
